File: modules/sql-backend/src/main/scala/RepositoryReader.scala
package edomata.backend

import cats.data.NonEmptyChain
import cats.effect.Concurrent
import cats.implicits.*
import edomata.core.*
import fs2.Pipe
import fs2.Stream

import java.time.OffsetDateTime
import java.util.UUID

enum AggregateState[S, E, R](val isValid: Boolean) {
  case Valid(state: S, version: SeqNr) extends AggregateState[S, E, R](true)
  case Conflicted(
      last: S,
      onEvent: EventMessage[E],
      errors: NonEmptyChain[R]
  ) extends AggregateState[S, E, R](false)
}

trait RepositoryReader[F[_], S, E, R] {
  def get(streamId: StreamId): F[AggregateState[S, E, R]]
  def history(streamId: StreamId): Stream[F, AggregateState[S, E, R]]
}

object RepositoryReader {
  def apply[F[_], S, E, R](
      journal: JournalReader[F, E],
      snapshot: SnapshotReader[F, S, E, R]
  )(using F: Concurrent[F], m: ModelTC[S, E, R]): RepositoryReader[F, S, E, R] =
    new {
      def get(streamId: StreamId): F[AggregateState[S, E, R]] =
        snapshot
          .get(streamId)
          .flatMap {
            case Some(last) =>
              journal
                .readStreamAfter(streamId, last.version)
                .through(scanState(last))
                .compile
                .lastOrError
            case None => history(streamId).compile.lastOrError
          }

      def history(streamId: StreamId): Stream[F, AggregateState[S, E, R]] =
        journal
          .readStream(streamId)
          .through(scanState(AggregateState.Valid(m.initial, 0L)))
    }
}

private[backend] def scanState[F[_], S, E, R](
    last: AggregateState[S, E, R]
)(using
    m: ModelTC[S, E, R]
): Pipe[F, EventMessage[E], AggregateState[S, E, R]] =
  _.scan(last) {
    case (AggregateState.Valid(s, version), ev) =>
      m.transition(ev.payload)(s)
        .fold(
          AggregateState.Conflicted(s, ev, _),
          AggregateState.Valid(_, version + 1)
        )
    case (other, ev) => other
  }.takeWhile(_.isValid, true)

File: rust/crates/edomata-backend/src/eventsourcing/repository_reader.rs
//! Folding the journal into aggregate states.

use std::sync::Arc;

use async_trait::async_trait;
use futures::stream::BoxStream;
use futures::{Stream, StreamExt, TryStreamExt};

use super::{AggregateState, JournalReader, SnapshotReader, ValidState};
use crate::{BackendError, EventMessage, Payload, SharedModel};

/// Reads the current state and the history of an aggregate.
#[async_trait]
pub trait RepositoryReader<S, E, R>: Send + Sync {
    /// Current state of the aggregate, folded from its snapshot and journal.
    async fn get(&self, stream_id: &str) -> Result<AggregateState<S, E, R>, BackendError>;

    /// Every state the aggregate went through, from the initial state to the
    /// current one (or to the first conflict, included).
    fn history(
        &self,
        stream_id: &str,
    ) -> BoxStream<'_, Result<AggregateState<S, E, R>, BackendError>>;
}

/// The default [`RepositoryReader`]: folds the journal with the model,
/// starting from the latest snapshot when there is one.
pub struct JournalRepositoryReader<S, E, R> {
    journal: Arc<dyn JournalReader<E>>,
    snapshot: Arc<dyn SnapshotReader<S>>,
    model: SharedModel<S, E, R>,
}

impl<S, E, R> JournalRepositoryReader<S, E, R> {
    /// Builds a reader over a journal and a snapshot reader.
    pub fn new(
        journal: Arc<dyn JournalReader<E>>,
        snapshot: Arc<dyn SnapshotReader<S>>,
        model: SharedModel<S, E, R>,
    ) -> Self {
        Self {
            journal,
            snapshot,
            model,
        }
    }
}

#[async_trait]
impl<S: Payload, E: Payload, R: Payload> RepositoryReader<S, E, R>
    for JournalRepositoryReader<S, E, R>
{
    async fn get(&self, stream_id: &str) -> Result<AggregateState<S, E, R>, BackendError> {
        match self.snapshot.get(stream_id).await? {
            Some(last) => {
                let events = self.journal.read_stream_after(stream_id, last.version - 1);
                last_state(scan_state(
                    events,
                    AggregateState::Valid(last),
                    &*self.model,
                ))
                .await
            }
            None => last_state(self.history(stream_id)).await,
        }
    }

    fn history(
        &self,
        stream_id: &str,
    ) -> BoxStream<'_, Result<AggregateState<S, E, R>, BackendError>> {
        let initial = AggregateState::Valid(ValidState::new(self.model.initial(), 0));
        Box::pin(scan_state(
            self.journal.read_stream(stream_id),
            initial,
            &*self.model,
        ))
    }
}

/// Emits `last`, then every state obtained by applying the events in order,
/// stopping after the first conflict (which is emitted).
pub(crate) fn scan_state<'a, S, E, R>(
    events: impl Stream<Item = Result<EventMessage<E>, BackendError>> + Send + 'a,
    last: AggregateState<S, E, R>,
    model: &'a (dyn edomata_core::DomainModel<State = S, Event = E, Rejection = R> + Send + Sync),
) -> impl Stream<Item = Result<AggregateState<S, E, R>, BackendError>> + Send + 'a
where
    S: Payload,
    E: Payload,
    R: Payload,
{
    let folded = events.scan(Some(last.clone()), move |current, event| {
        let next = match (current.take(), event) {
            (None, _) => None,
            (Some(_), Err(e)) => Some(Err(e)),
            (Some(AggregateState::Valid(ValidState { state, version })), Ok(ev)) => {
                match model.transition(&ev.payload, state.clone()) {
                    Ok(s) => {
                        let next = AggregateState::valid(s, version + 1);
                        *current = Some(next.clone());
                        Some(Ok(next))
                    }
                    Err(errors) => Some(Ok(AggregateState::Conflicted {
                        last: state,
                        on_event: ev,
                        errors,
                    })),
                }
            }
            (Some(conflicted @ AggregateState::Conflicted { .. }), Ok(_)) => {
                // Scala keeps emitting the conflicted state but `takeWhile`
                // stops right after it; we stop as well.
                let _ = conflicted;
                None
            }
        };
        std::future::ready(next)
    });
    futures::stream::once(std::future::ready(Ok(last))).chain(folded)
}

async fn last_state<S, E, R>(
    stream: impl Stream<Item = Result<AggregateState<S, E, R>, BackendError>> + Send,
) -> Result<AggregateState<S, E, R>, BackendError> {
    let mut stream = std::pin::pin!(stream);
    let mut last = None;
    while let Some(item) = stream.try_next().await? {
        last = Some(item);
    }
    last.ok_or_else(|| BackendError::persistence("empty aggregate history"))
}
